# Big Data: Tools and Methods

# Assignment 2 - Xin Zhou

<font color=blue>Requirement: In problems 1 - 4 you are to use higher order functions (map, reduce, filter, ...) to solve the problems. You do not need to use for or while statements. The solutions with map, reduce and/ or filter may not be as efficient as solutions using a for or while. However the core operations in Hadoop are map and reduce. While there is a difference in the Hadoop mapReduce from what you will use hear it is important to get used to using those constructs.
In each of the solutions to the problems you are to include tests that show that your code works.

<font color=green>The below function is a Test function for all problem sets, inputs are two arrays of test inputs for the test function and hand calculated results, it prints out the correct rate.

In [43]:
def test[T](inputs:Array[T], results:Array[T], f:T=>T){//can only test for 1 argument 1 result function
    if (inputs.length != results.length){throw new IllegalArgumentException("The dimensions of inputs mismatch") }
    var counts:Int = 0
    for ( (inputElement, trueResultElement) <- (inputs zip results)){
        var testResult = f(inputElement)
        if (testResult == trueResultElement){counts+=1}
        println("Input: "+inputElement+" | Output: "+testResult+" | Correct Output: "+trueResultElement)
    }
    println("Correct Rate: "+counts*100/inputs.length+"%")
}

defined function test

<font color=red>#1. <font color=green>If we list all the natural numbers below 20 that are multiples of 3 or 5 but not multiples both of 3 and 5, we get 3, 5, 6, 9, 10, 12, 18. The sum of these multiples is 63. Write a function, sum_multiples_3_5, that returns the sum of the multiples of 3 and 5 but not both less than N.

In [44]:
def sum_multiples_3_5(n:Int ): Int ={
  if (n < 0){throw new IllegalArgumentException("The input is not a natural number") }
  return (0 to n-1).toArray.filter(x => (x % 3 == 0 || x % 5 == 0) && x % 15 != 0).reduce(_ + _)
}

// Test for sum_multiples_3_5
//Input:16, Correct ouput: 3+5+6+9+10+12 = 45
//Input:20, Correct ouput: 3+5+6+9+10+12+18 = 63
//Input:25, Correct ouput: 3+5+6+9+10+12+18+20+21+24 = 128
//Input:31, Correct ouput: 3+5+6+9+10+12+18+20+21+24+25+27 = 180
test(Array(16,20,25,31), Array(45,63,128,180),sum_multiples_3_5)

Input: 16 | Output: 45 | Correct Output: 45
Input: 20 | Output: 63 | Correct Output: 63
Input: 25 | Output: 128 | Correct Output: 128
Input: 31 | Output: 180 | Correct Output: 180
Correct Rate: 100%


defined function sum_multiples_3_5

<font color=red>#2. <font color=green>Write a function, pattern_count with two arguments. The first arguments is a string, lets call it text, and the second argument is also a string, call it pattern. The function pattern_count returns the number of times the pattern occurs in the text. For example:<br>
pattern_count(“abababa”, “aba”) == 3 <br>
pattern_count(“aaaaa”, “aa”) == 4 <br> 
pattern_count(“Abcde”, “abc”) == 0 <br>

In [47]:
def pattern_count(text:String, pattern:String): Int= text.sliding(pattern.length).count(window => window == pattern)
//Test for pattern_count
pattern_count("abababa", "aba") //should return 3
pattern_count("aaaaa", "aa")    //should return 4
pattern_count("Abcde", "abc")   //should return 0

defined function pattern_count
res46_1: Int = 3
res46_2: Int = 4
res46_3: Int = 0

<font color=red>#3. <font color=green>Implement factorial using “to” and “reduceLeft” without a loop or recursion.

In [51]:
def factorial(n: Int): Int = {
    if (n < 1){throw new IllegalArgumentException("The input is not a natural number")}
    return (1 to n).product
}
//Test for factorial
//Input:5, Correct ouput: 1*2*3*4*5 = 120
//Input:6, Correct ouput: 1*2*3*4*5*6 = 720
//Input:7, Correct ouput: 1*2*3*4*5*6*7 = 5040
//Input:10, Correct ouput: 1*2*3*4*5*6*7*8*9*10 = 3628800
test(Array(5,6,7,10),Array(120,720,5040,3628800), factorial)
    

Input: 5 | Output: 120 | Correct Output: 120
Input: 6 | Output: 720 | Correct Output: 720
Input: 7 | Output: 5040 | Correct Output: 5040
Input: 10 | Output: 3628800 | Correct Output: 3628800
Correct Rate: 100%


defined function factorial

<font color=red>#4. <font color=green>Write a function that receives a collection of strings and a map from strings to integers. Return a collection of integers that are values of the map corresponding to one of the strings in the collection. For example, given Array("Tom", "Fred", "Harry") and Map("Tom" -> 3, "Dick" -> 4, "Harry" -> 5), return Array(3, 5).

In [17]:
def mapLookup(names: Array[String], dictionary: Map[String, Int]): Array[Int]=
dictionary.filterKeys(key => names.contains(key)).values.toArray
//Test for mapLookup
//Case 1, names: Array("Tom", "Fred", "Harry"), dictionary: Map("Tom" -> 3, "Dick" -> 4, "Harry" -> 5)
//Correct Output: Array(3, 5)
mapLookup(Array("Tom", "Fred", "Harry"), Map("Tom" -> 3, "Dick" -> 4, "Harry" -> 5))

//Case 2, names: Array("Dog", "Cat", "Bird", "Rabbit"), dictionary: Map("Cat" -> 1, 
//"Snake" -> 4, "Dog" -> 2, "Bird" ->3)
//Correct Output: Array(1, 2, 3)
mapLookup(Array("Dog", "Cat", "Bird", "Rabbit"), Map("Cat" -> 1, "Snake" -> 4, "Dog" -> 2, "Bird" ->3))

defined function mapLookup
res16_1: Array[Int] = Array(3, 5)
res16_2: Array[Int] = Array(1, 2, 3)

<font color=red>#5. <font color=green>Implement a function sampleNoRepeats with two arguments. The first argument is a range, the second an integer, n, greater that zero. The function returns n random numbers in the given range with no repeats.

In [1]:
def sampleNoRepeats(range: Range.Inclusive, n: Int): Array[Int]={
    if (n <= 0 || range.length < n){throw new IllegalArgumentException("Invalid input, the input number n has to be smaller or equal than the range") }
    val r = new scala.util.Random()
    return r.shuffle(range).toArray.take(n)
}
//Test for sampleNoRepeats
//sampleNoRepeats(5 to 20, 4), should be in range of 5 to 20, no repeats
//sampleNoRepeats(10 to 20, 4), should be in range of 10 to 20, no repeats
//sampleNoRepeats(10 to 20, 10), should be in range of 10 to 20, no repeats
sampleNoRepeats(5 to 20, 4)
sampleNoRepeats(10 to 20, 4)
sampleNoRepeats(10 to 20, 10)

defined function sampleNoRepeats
res0_1: Array[Int] = Array(7, 15, 6, 9)
res0_2: Array[Int] = Array(14, 11, 13, 17)
res0_3: Array[Int] = Array(11, 18, 15, 13, 12, 10, 14, 17, 20, 19)

<font color=red>Question 6~9
#6. <font color=green>Create an array with random 100,000 integers between 0 and 50,000. Compute the mean value of the array, call it M. Take a random sample of 100 different elements (sampling without replacement) of the array and compute the mean of the sample. What is the error?
<font color=red>#7. <font color=green>Using the same array as in#6 take 1,000 samples of 100 random elements of the array. Compute the mean of each sample. Find the value r where interval (M-r, M+r) contains 95% of the means of the 1000 samples.
<font color=red>#8. <font color=green>Repeat #7 but use the sample size of 1,000 rather than 100.
<font color=red>#9. <font color=green>Repeat #8 but use the sample size of 10,000 rather than 100.

In [13]:
//#6
def random_Multi_Ints(start:Int, end:Int, counts:Int):(Array[Int], Double)={
    val r = new scala.util.Random
    val list =  Array.fill(counts){start + r.nextInt( (end - start) + 1 )}
    val mean = list.map(_.toDouble/counts).reduce(_ + _)
    return (list, mean)
}
val (randomInts, mean100k) = random_Multi_Ints(0, 50000, 100000);
// println("randomInts has "+randomInts.length+" values") //check if the output is desired size of array, 100000 values
// println("randomInts maximum value: "+randomInts.max)  //check if the random values are within range, min >= 0
// println("randomInts minumum value: "+randomInts.min)  //check if the random values are within range, max <= 50000

def random_Sample_Mean(data: Array[Int], size: Int): Double={
    val r = new scala.util.Random()
    var list = (r.shuffle(0 to data.length-1).take(size) map data).toArray
//     println("created list" + size+" has "+list.length+" values") //check if the output is desired size of array, 100 values
//     println("created list" + size+" maximum value: "+list.max)  //check if the random values are within range, min >= 0
//     println("created list" + size+" minumum value: "+list.min)  //check if the random values are within range, max <= 50000
    return list.map(_.toDouble/size).reduce(_ + _)
}
var sample100_Mean = random_Sample_Mean(randomInts, 100);
var error = math.round(math.abs(mean100k-sample100_Mean))
var errorRate = math.round(error*10000/mean100k).toDouble/100
println("The mean of sample of 100 numbers comparing to the mean of whole data has error of "+
        error+", the error rate is "+errorRate+"%")



The mean of sample of 100 numbers comparing to the mean of whole data has error of 477, the error rate is 1.91%


defined function random_Multi_Ints
randomInts: Array[Int] = Array(
  21116,
  26296,
  34478,
  14328,
  42839,
  5137,
  32865,
  32607,
  11926,
  27004,
  17248,
...
mean100k: Double = 25012.259710000457
defined function random_Sample_Mean
sample100_Mean: Double = 24535.400000000005
error: Long = 477L
errorRate: Double = 1.91

In [15]:
//#7~9
//In a normal distributed random values, For a 95.45% of samples, it has a interval of (μ - 2σ < x < μ + 2σ)
//finding the standard deviation would have the range which is 2σ
def twoSigma(data: Array[Int], size: Int, numOfSamples: Int): Int={
    val means = Array.fill(numOfSamples){random_Sample_Mean(data, size)}
    val meanOfMeans = means.map(_.toDouble/numOfSamples).reduce(_ + _)
    val devs = means.map(eachMean => (eachMean - meanOfMeans) * (eachMean - meanOfMeans))
    val stddev = Math.sqrt(devs.map(_.toDouble/(numOfSamples-1)).reduce(_ + _))
    return (2*stddev).toInt
}
val r7 = twoSigma(randomInts, 100, 1000) //sample size of 100, 1000 samples
val r8 = twoSigma(randomInts, 1000, 1000) //sample size of 1000, 1000 samples
val r9 = twoSigma(randomInts, 10000, 1000) //sample size of 10000, 1000 samples


defined function twoSigma
r7: Int = 2866
r8: Int = 930
r9: Int = 271

Results: <br>
--------------
r7 = 2866<br>
r8 = 930<br>
r9 = 271<br>
Comment:<br>
--------------
We can tell that as we increase the size of samples of means, the range become smaller and smaller closing the average of means.

In [1]:
var data=Array(1.2,2.3,4.5,6.7)

data: Array[Double] = Array(1.2, 2.3, 4.5, 6.7)